sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -73,56 +170,187 @@ fn node_name(subsystem_vid: i32, subsystem_id: i32) -> String {
+    match slot {
+        0x00..=0x09 => Some(DiskVariant::U2),
+        0x11..=0x12 => Some(DiskVariant::M2),
+        _ => None,
+    }
+}
+
+fn poll_tofino_node(
+    log: &Logger,
+    tofino: &mut TofinoSnapshot,
+    node: &Node<'_>,
+) {
+    if node.node_name() == node_name(TOFINO_SUBSYSTEM_VID, TOFINO_SUBSYSTEM_ID)
+    {
+        tofino.exists = true;
+        tofino.driver_loaded = node.driver_name().as_deref() == Some("tofino");
+        debug!(
+            log,
+            "Found tofino node, with driver {}loaded",
+            if tofino.driver_loaded { "" } else { "not " }
+        );
+    }
+}
+
+fn get_dev_path_of_whole_disk(
+    node: &Node<'_>,
+) -> Result<Option<PathBuf>, Error> {
+    let mut wm = node.minors();
+    while let Some(m) = wm.next().transpose().map_err(Error::DevInfo)? {
+        // "wd" stands for "whole disk"
+        if m.name() != "wd" {
+            continue;
+        }
+        let links = {
+            match DevLinks::new(true) {
+                Ok(links) => links,
+                Err(_) => DevLinks::new(false).map_err(Error::DevInfo)?,
+            }
+        };
+        let devfs_path = m.devfs_path().map_err(Error::DevInfo)?;
+
+        let paths = links
+            .links_for_path(&devfs_path)
+            .map_err(Error::DevInfo)?
+            .into_iter()
+            .filter(|l| {
+                // Devices in "/dev/dsk" have names that denote their purpose,
+                // of the form "controller, disk, slice" or "controller, disk,
+                // partition".
+                //
+                // The suffix of "d0" is typical of an individual disk, and is
+                // the expected device to correspond with the "wd" device in
+                // the "/devices" hierarchy.
+                l.linktype() == DevLinkType::Primary
+                    && l.path()
+                        .file_name()
+                        .map(|f| f.to_string_lossy().ends_with("d0"))
+                        .unwrap_or(false)
+            })
+            .collect::<Vec<_>>();
+
+        if paths.is_empty() {
+            return Err(Error::NoDevLinks(PathBuf::from(devfs_path)));
+        }
+        return Ok(Some(paths[0].path().to_path_buf()));
+    }
+    Ok(None)
+}
+
+// Gather a notion of "device identity".
+fn blkdev_device_identity(devfs_path: &str) -> Result<DiskIdentity, Error> {
+    let path = PathBuf::from(devfs_path);
+    let ctrl_id = nvme::identify_controller(&path)
+        .map_err(|err| Error::MissingNVMEIdentity { path, err })?;
+
+    let vendor =
+        format!("{:x}{:x}", ctrl_id.vendor_id, ctrl_id.subsystem_vendor_id);
+    let serial = ctrl_id.serial;
+    let model = ctrl_id.model;
+    Ok(DiskIdentity { vendor, serial, model })
+}
+
+fn poll_blkdev_node(
+    log: &Logger,
+    disks: &mut HashSet<UnparsedDisk>,
+    mut node: Node<'_>,
+) -> Result<(), Error> {
+    if let Some(driver_name) = node.driver_name() {
```

Comment:
Could use the new `let else` syntax here to reduce rightward drift, I think? Something like

```rust
let Some(driver_name) = node.driver.name() else {
    return Ok(());
};
```

I'd probably also throw in

```rust
if driver_name != "blkdev" {
    return Ok(());
}
```

to shift this whole function over a couple levels.

Reply:
Good catch, updated!